app/scripts/controllers/permissions/index.js

```diff
@@ -0,0 +1,418 @@
+   *
+   * @param {string} origin - The origin to change the exposed accounts for.
+   * @param {string[]} accounts - The new account(s) to expose.
+   */
+  async updateExposedAccounts (origin, accounts) {
+
+    await this.validateExposedAccounts(accounts)
+
+    this.permissions.updateCaveatFor(
+      origin, 'eth_accounts', CAVEAT_NAMES.exposedAccounts, accounts
+    )
+
+    this.notifyDomain(origin, {
+      method: ACCOUNTS_CHANGED_NOTIFICATION,
+      result: accounts,
+    })
+  }
+
+  /**
+   * Finalizes a permissions request.
+   * Throws if request validation fails.
+   *
+   * @param {Object} requestedPermissions - The requested permissions.
+   * @param {string[]} accounts - The accounts to expose, if any.
+   */
+  async finalizePermissionsRequest (requestedPermissions, accounts) {
+
+    const { eth_accounts: ethAccounts } = requestedPermissions
+
+    if (ethAccounts) {
+
+      await this.validateExposedAccounts(accounts)
+
+      if (!ethAccounts.caveats) {
+        ethAccounts.caveats = []
+      }
+
+      // caveat names are unique, and we will only construct this caveat here
+      ethAccounts.caveats = ethAccounts.caveats.filter(c => (
+        c.name !== CAVEAT_NAMES.exposedAccounts
+      ))
+
+      ethAccounts.caveats.push(
+        {
+          type: 'filterResponse',
+          value: accounts,
+          name: CAVEAT_NAMES.exposedAccounts,
+        },
+      )
+    }
+  }
+
+  /**
+   * Validate an array of accounts representing accounts to be exposed
+   * to a domain. Throws error if validation fails.
+   *
+   * @param {string[]} accounts - An array of addresses.
+   */
+  async validateExposedAccounts (accounts) {
+
+    if (!Array.isArray(accounts) || accounts.length === 0) {
+      throw new Error('Must provide non-empty array of account(s).')
+    }
+
+    // assert accounts exist
+    const allAccounts = await this.keyringController.getAccounts()
+    accounts.forEach(acc => {
+      if (!allAccounts.includes(acc)) {
+        throw new Error(`Unknown account: ${acc}`)
+      }
+    })
+  }
+
+  /**
+   * Removes the given permissions for the given domain.
+   * @param {object} domains { origin: [permissions] }
+   */
+  removePermissionsFor (domains) {
+
+    Object.entries(domains).forEach(([origin, perms]) => {
+
+      this.permissions.removePermissionsFor(
+        origin,
+        perms.map(methodName => {
+
+          if (methodName === 'eth_accounts') {
+            this.notifyDomain(
+              origin,
+              { method: ACCOUNTS_CHANGED_NOTIFICATION, result: [] }
+            )
+          }
+
+          return { parentCapability: methodName }
+        })
+      )
+    })
+  }
+
+  /**
+   * Gets all caveats for the given origin and permission, or returns null
+   * if none exist.
+   *
+   * @param {string} permission - The name of the target permission.
+   * @param {string} origin - The origin that has the permission.
+   */
+  getCaveatsFor (permission, origin) {
```

Comment:
These two methods (`getCaveat` and `getCaveatsFor`) are both unused at the moment.